core/src/main/java/com/linecorp/armeria/internal/metric/PrometheusRegistry.java

```diff
@@ -0,0 +1,207 @@
+     * @return existing or newly created {@link Collector}
+     * @throws IllegalArgumentException if one of the following conditions meet:
+     * <ul><li>A {@link Collector} with the same metric family name was registered to the
+     * {@code collectorRegistry} without going through this method. e.g. <pre>{@code
+     * CollectorRegistry registry = new CollectorRegistry();
+     * // Register without going through the wrapper.
+     * registry.register(Counter.build("foo", "counting the number of foos"));
+     * // Will fail with an IllegalArgumentException.
+     * registerIfAbsent(registry, "foo", name -> Counter.build(name, "counting the number of foos"));
+     * }</pre></li>
+     * <li>One of the samples produced by an existing {@link Collector} has the same name. e.g. <pre>{@code
+     * registerIfAbsent(registry, "foo", name -> Summary.build("foo", "the summary of foos"));
+     * // Will fail because the summary above will register "foo_count".
+     * registryIfAbsent(registry, "foo_count", name -> Counter.build(name, "counting the number of foos"));
+     * }</pre></li></ul>
+     */
+    static <T extends Collector> T registerIfAbsent(
+            CollectorRegistry collectorRegistry, String name, Function<String, T> collectorFactory) {
+        requireNonNull(collectorRegistry, "collectorRegistry");
+        requireNonNull(name, "name");
+        requireNonNull(collectorFactory, "collectorFactory");
+
+        checkArgument(METRIC_NAME_PATTERN.matcher(name).matches(),
+                      "name: %s (expected: matches %s)", name, METRIC_NAME_PATTERN);
+
+        synchronized (collectors) {
+            final Map<String, Collector> map =
+                    collectors.computeIfAbsent(collectorRegistry, unused -> new HashMap<>());
+
+            Collector collector = map.get(name);
+            if (collector == null) {
+                // Create a new Collector using the given factory.
+                collector = collectorFactory.apply(name);
+                requireNonNull(collector, "collectorFactory returned null.");
+
+                // Get the metric family name of the collector.
+                final List<MetricFamilySamples> mfsList = describe(collector);
+                final List<String> actualNames = new ArrayList<>(mfsList.size());
+                boolean containsMatchingName = false;
+                for (MetricFamilySamples mfs : mfsList) {
+                    if (mfs == null) {
+                        continue;
+                    }
+
+                    final String actualName = mfs.name;
+                    if (actualName == null) {
+                        continue;
+                    }
+
+                    actualNames.add(actualName);
+                    containsMatchingName = actualName.equals(name);
+                }
+
+                // Make sure the collector actually produced the metric with the specified metric family name.
+                if (!containsMatchingName) {
```

Comment:
I'm not sure, but this value depends on the last call of `containsMatchingName = actualName.equals(name)` of line 107? Is it correct?